### ISSUE TITLE
Add `channels_last` argument to `transforms.ToTensor()`

### ISSUE THREAD

**Issue by netw0rkf10w** [pytorch/vision#7090]
### 🚀 The feature

Currently, for using `channels_last`:

- During data loading, `torchvision.transforms.ToTensor()` converts the input images into CHW.
- Then during training, the images are converted **back** to HWC: `images = images.to(memory_format=torch.channels_last)`.

This is highly inefficient. I suggest to add an optional boolean argument to `ToTensor()` that skips the permutation and outputs directly HWC:

`torchvision.transforms.ToTensor(channels_last=True)`

### Motivation, pitch

To avoid unnecessary axis permutations and improve efficiency.

### Alternatives

I had to write my custom `functional.to_tensor` to add the suggested argument.

**Comment by pmeier** [comment]
> * Then during training, the images are converted **back** to HWC: `images = images.to(memory_format=torch.channels_last)`.

That is not what is happening. `memory_format=torch.channels_last` changes the way the tensor is laid out in memory:

```pycon
>>> t = torch.empty(2, 3, 4, 5)
>>> t.shape
torch.Size([2, 3, 4, 5])
>>> t.stride()
(60, 20, 5, 1)
>>> t = t.to(memory_format=torch.channels_last)
>>> t.shape
torch.Size([2, 3, 4, 5])
>>> t.stride()
(60, 1, 15, 3)
```

As you can see, the `shape` is exactly the same, but the stride is different.

**Comment by netw0rkf10w** [comment]
@pmeier Thanks. It turns out I have misunderstood how `to(memory_format=torch.channels_last)` works. 
Though it seems to me that the current `channels_last` pipeline remains suboptimal:

1. Is `t.to(memory_format=torch.channels_last)` expensive? Recall that we need to to this at every iteration, unlike for the model.
2. Is there a way to make the model accept NHWC inputs? It doesn't sound logical to convert the input images to CHW in the first place, if we are using the `channels_last` format.

**Comment by NicolasHug** [comment]
> Is t.to(memory_format=torch.channels_last) expensive?

It is a no-op at best (if `t` is already channels_last) and involves a copy of `t` otherwise. As @pmeier noted this will *never* change the shape of a tensor, only its internal representation / layout.

IMO the original request is reasonable : ToTensor() shouldn't convert strides on behalf of the user, considering how impactful strides can be w.r.t. performance performance of a pipeline. Could be something to consider in the revamp?

**Comment by NicolasHug** [comment]
@pmeier  good to know. I had some questions about this deprecation, let's discuss more offline.

@netw0rkf10w summarizing the situation:

Tensor shape and contiguity (as defined by the strides) are different things. A tensor can be of shape NCHW while being represented as a NHWC tensor internally (channels_last) or NCHW (channels_first), or any other exotic layout.

All models / transforms / operators in pytorch / torchvision except the *shape* to be NCHW, but the contiguity can be arbitrary. Most operators will try to preserve the contiguity of the input, but this may vary. A lot of operators are sensitive to contiguity when it comes to perf.

`to.(memory_format=)` is cheap in itself, but because operators are sensitive to contiguity, this can have effect on perf on the whole transforms pipeline. This isn't something easy to control in torchvision right now, but something we're looking into

**Comment by netw0rkf10w** [comment]
> > Recall that we need to to this at every iteration, unlike for the model.
> 
> Could you clarify what "we" means here? `grep`'ing for `channels_last` yields no results in this whole repository.

Sorry I was not very clear. I meant the user has to convert every input passed to the model:

```python
# Need to be done once, after model initialization (or load)
model = model.to(memory_format=torch.channels_last)

for input, label in dataloader:
    # Need to be done for every input
    input  = input.to(memory_format=torch.channels_last)
    output = model(input)
    ...
```

**Comment by pmeier** [comment]
> Sorry I was not very clear. I meant the user has to convert every input passed to the model:

No, you don't have to do that. The memory format is somewhat an implementation detail and if you don't want to you don't need to care about it. Changing the memory layout, i.e.

```python
input  = input.to(memory_format=torch.channels_last)
```

is only needed for performance reasons. As @NicolasHug stated above, some operators are very sensitive and quite a bit faster in different formats. But, to repeat myself, unless you are specifically looking into the performance, you don't need to touch it.

**Comment by netw0rkf10w** [comment]
@pmeier Thanks. But of course I am talking about performance. Isn't that the whole point of `channels_last` in the first place?

**Comment by netw0rkf10w** [comment]
By the way, it seems to me that the `F.to_tensor()` function:
 https://github.com/pytorch/vision/blob/8985b598a69250d65959941c863d76a4225ae7ac/torchvision/transforms/functional.py#L125 

can be modified as follows to accept the `channels_last` argument:


```
def to_tensor(pic, channels_last=False) -> Tensor:
    ...
    if isinstance(pic, np.ndarray):
        ...
        img = torch.from_numpy(pic.transpose((2, 0, 1)))
        if not channels_last:
            img = img.contiguous()
        ...

    ...
    # put it from HWC to CHW format
    img = img.permute((2, 0, 1))
    if not channels_last:
        img = img.contiguous()
    ...
```

Now, if we use `ToTensor(channels_last=True)` for the data loader, then we don't need to convert the input anymore:

```python
# Need to be done once, after model initialization (or load)
model = model.to(memory_format=torch.channels_last)

for input, label in dataloader:
    # No need, as the input is already the channels_last format
    # input  = input.to(memory_format=torch.channels_last)
    output = model(input)
    ...
```

**Comment by pmeier** [comment]
> But of course I am talking about performance. Isn't that the whole point of `channels_last` in the first place?

Yeah, but given the misunderstanding in the beginning, I didn't get that this was your point. 

Regarding the feature itself, I'm not sure that putting it on `ToTensor()` makes much sense: `t.to(memory_format=torch.channels_last)` requires a 4D tensor:

```pycon
>>> t = torch.empty(3, 224, 224)
>>> t.to(memory_format=torch.channels_last)
RuntimeError: required rank 4 tensor to use channels_last format
>>> t.unsqueeze(0).to(memory_format=torch.channels_last)
[...]
```

 The canonical way of using the transformations is through any of the `transform` arguments on the datasets:

```python
transform = transforms.Compose([transforms.ToTensor(channels_last=True), ...])
dataset = datasets.ImageNet(..., transform=transform)
```

But the dataset are going to return only a single image, i.e. a 3D tensor. Meaning, you could only use the `ToTensor(channels_last=True)` transformation *after* the data loader so basically as first step in the data loop. And at this point you need to have used `ToTensor` already, because you can't batch PIL images.

In total, your snippet would change from

```python
model = model.to(memory_format=torch.channels_last)

for input, label in dataloader:
    input  = input.to(memory_format=torch.channels_last)
    output = model(input)
    ...
```

to

```python
model = model.to(memory_format=torch.channels_last)

to_channels_last = ToTensor(channels_last=True)

for input, label in dataloader:
    input  = to_channels_last(input)
    output = model(input)
    ...
```

and I don't see the appeal in that.

**Comment by pmeier** [comment]
I've just checked that even if we fake the `channels_last` layout by unsqueezing -> converting -> squeezing, which in our past benchmarks also benefited from the speed-ups, the stacking in the data loader makes everything contiguous again:

```python
import torch

images = [torch.zeros(3, 224, 224), torch.ones(3, 224, 224)]
images = [image.unsqueeze(0).to(memory_format=torch.channels_last).squeeze(0) for image in images]
batch = torch.stack(images, 0)
assert batch.is_contiguous()
```
